src/vs/workbench/contrib/terminal/browser/terminalService.ts

```diff
@@ -651,24 +651,13 @@ export class TerminalService implements ITerminalService {
 	private _onWillShutdown(e: WillShutdownEvent): void {
 		// Don't touch processes if the shutdown was a result of reload as they will be reattached
 		const shouldPersistTerminals = this._configHelper.config.enablePersistentSessions && e.reason === ShutdownReason.RELOAD;
-		const terminals = [...this._terminalGroupService.instances, ...this._backgroundedTerminalInstances];
-
-		// Dispose of all transient terminals
-		for (const instance of terminals.filter(i => !i.shouldPersist)) {
-			instance.dispose(TerminalExitReason.Shutdown);
-		}
 
-		// Detach all persistent terminals
-		if (shouldPersistTerminals) {
-			for (const instance of terminals.filter(i => i.shouldPersist)) {
+		for (const instance of [...this._terminalGroupService.instances, ...this._backgroundedTerminalInstances]) {
+			if (shouldPersistTerminals && instance.shouldPersist) {
 				instance.detachProcessAndDispose(TerminalExitReason.Shutdown);
+			} else {
+				instance.dispose(TerminalExitReason.Shutdown);
 			}
-			return;
-		}
-
-		// Force dispose of all pane terminal instances as they will be recreated on restart
-		for (const instance of this._terminalGroupService.instances) {
-			instance.dispose(TerminalExitReason.Shutdown);
 		}
 
 		// Clear terminal layout info only when not persisting
```

Comment:
We only want to do this when `!shouldPersistTerminals`

Reply:
we also want to do that when `!instance.shouldPersist`

Reply:
I mean the `// Clear terminal layout info only when not persisting` part